### ISSUE TITLE
solid-styled-components: TypeScript - Base tag props not exposed

### ISSUE THREAD

**Issue by kavsingh** [ryansolid/solid#147]
## Overview
When wrapping with styled, wrapped tag props don't seem to be exposed.

---

Example repo: https://github.com/kavsingh/tsblank/tree/solid-js-styled
`npm i` and open `src/components/count.tsx` with type linting enabled:

![image](https://user-images.githubusercontent.com/813450/80211931-e2398d80-8636-11ea-941d-5010bb0eb149.png)

Thanks!

**Comment by ryansolid** [comment]
Ok, my TypeScript isn't always the best. If someone is interested in submitting a PR that would be appreciated. Otherwise, I will look into this weekend.

**Comment by high1** [comment]
I took a look at this - in
```
export function styled<T = {}>
```
this should be extended with HTMLAttributes 
```
export function styled<T = {} & HTMLAttributes<unknown>>
```
to cover most of the base cases - but it's not complete. React definitions for JSX have AllHTMLAttributes interface, but I don't see something resemblingthat  in JSX definitions for Solid. I guess that something in line of 
```
export function styled<T = {} & AllHTMLAttributes<unknown>>
```
would fit for this case?

**Comment by high1** [comment]
I took a look at this, and then at new goober typings - and came up with something like this:
```
import { css, CSSAttribute } from 'goober';
import { cloneProps, createContext, useContext } from 'solid-js';
import { spread } from 'solid-js/dom';

const ThemeContext = createContext();

type StyledTemplateArgs<T> = [
  CSSAttribute | TemplateStringsArray | string | ((props: T) => CSSAttribute | string),
  ...Array<string | number | ((props: T) => CSSAttribute | string | number | undefined)>
];

export function styled<T extends keyof JSX.IntrinsicElements>(tag: T) {
  return <P extends {}>(...args: StyledTemplateArgs<T & P & { theme?: any }>) => {
    return (
      props: P & JSX.IntrinsicElements[T] & { className?: string; children?: any; theme?: any }
    ): JSX.IntrinsicElements[T]=> {
      const newProps = cloneProps(props);
      props.theme = useContext(ThemeContext);
      Object.defineProperty(newProps, 'className', {
        get() {
          const pClassName = props.className,
            append = 'className' in props && /^go[0-9]+/.test(pClassName!);

          // Call `css` with the append flag and pass the props
          const className = css.apply(
            { target: this.target, o: append, p: props },
            args as [any, ...any[]]
          );

          return [pClassName, className].filter(Boolean).join(' ');
        },
        configurable: true,
        enumerable: true,
      });

      const el = document.createElement(tag);
      spread(el, newProps);

      return el;
    };
  };
}
```
Usage:
```
interface ButtonProps {
  height?: number;
}

const TestDiv = styled('div')<ButtonProps>`
  border-radius: 20px;
  height: ${({ height = 100 }) => height}px;
  background: red;
`;

const TestAnchor = styled('a')`
  background: blue;
`;

<TestDiv onClick={() => alert('test')} />
<TestAnchor href="#" />
```
Type of TestDiv is JSX.HTMLAttributes<HTMLDivElement>, I'm not sure if that;s quite correct. But it worked when I tested it here. Now, the string for tag is type checked, so you cant call it like styled('xyz'). Also, it should pick up props from the JSX.IntrinsicElements[T], so assigning href to TestDiv would fail.

**Comment by ryansolid** [comment]
Hmmm.. I guess that checks because from a JSX perspective it's only ever the attributes that are being looked at. This whole area seems so complicated. Once I get the JSX types in Solid I suppose this will be easier to wire up. In general, I'm so confused over Components even referencing JSX types when they don't need to use JSX. Preact often uses h directly or htm these days. Are JSX types throughout their type declarations. ... And yes that seems to be the case. I mean sure they reference their element type in most places but when it comes down to intrinsic elements. I guess I shouldn't worry about this at all given the types will be shipped with Solid.

**Comment by high1** [comment]
Looking at the new goober typings for styled, i see that they've imported React typings:
```
interface StyledFunction {
        // used when creating a styled component from a native HTML element
        <T extends keyof JSX.IntrinsicElements, P extends Object = {}>(tag: T): Tagged<
            JSX.LibraryManagedAttributes<T, JSX.IntrinsicElements[T]> & P
        >;

        // used to extend other styled components. Inherits props from the extended component
        <PP extends Object = {}, P extends Object = {}>(tag: StyledVNode<PP>): Tagged<PP & P>;

        // used when creating a component from a string (html native) but using a non HTML standard
        // component, such as when you want to style web components
        <P extends Object = {}>(tag: string): Tagged<P & Partial<JSX.ElementChildrenAttribute>>;

        // used to create a styled component from a JSX element (both functional and class-based)
        <T extends JSX.Element | JSX.ElementClass, P extends Object = {}>(tag: T): Tagged<P>;
    }
```
My implementation covers the first case, and I've omitted JSX.LibraryManagedAttributes because I guess there's no default props in Solid like in React.
https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-0.html#support-for-defaultprops-in-js
Looking at this, I see that there are 3 more cases, but I'd be more concerned that goober now  has global JSX namespace due to React typings import.

**Comment by high1** [comment]
Excuse my ignorance, but I saw multiple times that Babel is the only option for doing the code optimization. If you would introduce a JSXfactory f- a function that wold render DOM elements from JSX, couldn't it also be an option? The reason why I'm asking is - if that is possible, this could largely increase compatibility with existing libraries - for instance, goober setup could be used with this function to make goober work correctly, and thus eliminate the need for a custom wrapper. Also, global JSX export would not be needed. But again, it's just an uneducated guess...

**Comment by ryansolid** [comment]
Hey I do have a JSXFactory function in `solid-js/h` which is there purely for compatibility since I have a more performant non-compiled approach in `solid-js/html`. So it's doable, I just don't make that the main solution since it can't be optimized in the same way. Basically the big benefit here is the DOM Expression's compiler. Without it you might as well be using a Virtual DOM since you have to do all the same work, and put a reactive system on top. The compiler is unique angle that reactive libraries can take to remove the guesswork. An HyperScript factory always needs to ask the question with new data of what it's doing. It doesn't have the ability to create DOM nodes on mass. It requires manual indication of what's reactive or not. It has to do all these checks at runtime and at a per element granularity. Atleast Tagged Template Template literals (`solid-js/html`) can analyse the whole template and cache the execution. HyperScript executes inside out:

```js
// which h function executes first?
h("div", { id: "main" }, [
  h("div", "hello"),
  h("div", h("span", () => state.name))
])
```
This lends to creating the structure first then walking over it again to build something. Now you could probably find a way using wrappers to cache this information similar to Tagged Template Literals after the first execution (and I don't mean reactive execution since that never runs again, more like the next row). But it requires new wrappers syntax that would break compatibility anyway. 

I think this is what Rich Harris means when he says Svelte can't use JSX. He isn't talking about JSX, he means HyperScript. I have gone through the effort of making Solid HyperScript compatible and it is pretty performant but it just can't be the de facto approach. See my article [The Fastest Way to Render the DOM](https://medium.com/better-programming/the-fastest-way-to-render-the-dom-e3b226b15ca3?source=friends_link&sk=5ae1688dde789e46cecf5c976e708da5).

**Comment by high1** [comment]
So there so possibility of a hybrid way out of of this? To make this JSX factory and the compilation step work together, and exchange information? The goal here is to reuse the previous work - because if this is not a possibility, I fear that there will be an issue raised  for each popular JSX library going forwared. Taking a thorough read of the article, thanks.

**Comment by ryansolid** [comment]
I mean there is a hybrid. Sort of if you ignore reactivity. But you can't really ignore reactivity. In some cases it will just work but it will never be the optimal approach. I guess I never viewed things as JSX libraries but rather React libraries or React-like libraries. Most things in this space are never going to work simply because they expect a React like update cycle regardless of what the syntax is. Its the same reason Solid will never be a swap in for Function Components with hooks. The update cycle is backwards.

Goober does attempt to make a general JSX approach but it is for the most part just using HyperScript. HyperScript is innately VDOM. It basically constructs a virtual tree. Stuff like the way it executes inside out doesn't lend to single pass approaches.  JSX on the other hand is this: https://facebook.github.io/jsx/. No more not less. People make HyperScript libraries not JSX libraries from my perspective because they work without JSX. HyperScript is just the most common JSX target.

**Comment by high1** [comment]
Agreed. Unfortunately, expectations go beyond that... OK,I guess we'll have to wait for thing to unwind.

**Comment by ryansolid** [comment]
Yeah I guess you can appreciate more why I realized I needed to make my own library wrappers. I sort of didn't want to take it on, but realistically I need some projects out there so people can get started. But there will be gaps at first, especially when I'm implementing things that I don't have a lot of experience with. It's a bit of a slog as there are always a couple cases I miss. Which is largely why I haven't been comfortable dropping 1.0 yet. But as more people have been using and reporting issues the "solid" this is all getting.

**Comment by high1** [comment]
Yeah, noted... That's quite a lot take on - and it's not the way forward. It's just not doable. I mean, Iget the impression that it would be easier to rewrite goober's styled function than to write typed compatible implementations in TypeScript.

**Comment by ryansolid** [comment]
I believe I've done something reasonable here based on @high1 suggestions. As far as I can tell this worked in my tests. Please reopen if still an issue. I'm happy to report this is fixed in v0.18.0.
